### DISCUSSION TITLE
[FeatureRequest] Improve visualization settings to specify the source of Labels and Values

### DISCUSSION THREAD

**Discussion by Totalus** [grafana/grafana#38277] [Dashboards, Panels & Visualizations]
### Why do we need this ?

Currently some visualizations (such as Pie Chart and Bar Gauge) seem to either use field names or series names as labels, but one may have data structured like this:

| City | Population (thousands) | Temperature |
|--|--|--
| London | 9304 | 16°C
| New York | 8462 | 19°C
| Montreal | 2070 | 25°C

And would like to make a Bar Gauge graph with population from each city, but it does not seem possible to do that visualization when the data is structured this way. The *City* values would need to be the field names or series name instead to be used as labels by the visualization.

I first thought about a new transformation that could help with that (#33824), but then also thought about the possibility to handle this in the visualization configuration directly (this issue).

### What would you like to be added ?

Modify the visualization display settings so we can specify more precisely how to use the input data to construct the charts. We can brainstorm about the possibilities. The different scenarios I can se are:

1. Use field names as labels + select a reduce function for the values
2. Use series names as labels + select which fields to use as values + select reduce function
3. Use specified field rows as labels + select which other field rows to use for values (currently not possible)

The two first seem to be the default behaviour, but the third is not possible (or I don't know how to do it). We could add some configuration inputs to handle this case. For instance we could add an input to first select what is the source of the Labels (Field name / Series name / Field value), then depending on the selection, either select the reduce function, or the Field name to use for Labels and for Values, or some kind of combination like that, we can brainstorm further on this.

This would allow to use the above table structure as input data while also making it clearer for the user I believe. Currently the way those visualizations use the input data is not much documented and a bit unintuitive. This may also apply to other visualizations as well.

**Comment by hugohaggmark** [comment]
@ryantxu @torkelo does this look promising?

**Comment by torkelo** [comment]
This is supported, need to show all values under value options. 

Grafana will use the string field as the name for each value.

**Comment by torkelo** [comment]
And the new bar chart in v8 only supports this form of data

**Comment by Totalus** [comment]
> This is supported, need to show all values under value options.
> Grafana will use the string field as the name for each value.

Ah yes you're right, the real issue is with the naming of each value, it is all the same (the column name) which makes basically unusable.

![image](https://user-images.githubusercontent.com/34516402/121030241-4eae1800-c777-11eb-89eb-e8c3674c2ef2.png)

![image](https://user-images.githubusercontent.com/34516402/121031263-3094e780-c778-11eb-92b4-1787f336b6af.png)

![image](https://user-images.githubusercontent.com/34516402/121032350-30e1b280-c779-11eb-80da-354c9ec6e1bd.png)


> And the new bar chart in v8 only supports this form of data

Seems like v8 handles the naming better indeed.

**Comment by torkelo** [comment]
yea with that data in v8 Grafana should name each value according to the value in the same row in the string field

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
